In [2]:
# import libaries
import pandas as pd
import numpy as np

# import visualization tools
import matplotlib.pyplot as plt
import seaborn as sns

# import machine learning libaries
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier, export_graphviz
from sklearn.metrics import classification_report
from sklearn import tree

# import NLP tool
from wordcloud import WordCloud

# import self-created functions
import prepare
import project_acquire
import explore

# Wrangle

In [7]:
# use prepare module wrangle data function to acquire data
data=pd.read_json('data.json')

In [15]:
# check if we successfully aquire the data and clean it
data.readme_contents[10]

'<center>\n    <img style="border-radius: 0.3125em;\n    box-shadow: 0 2px 4px 0 rgba(34,36,38,.12),0 2px 10px 0 rgba(34,36,38,.08);" \n    src="./data/.logo图片/.img.jpg"width="180">\n    <br>\n    <div style="color:orange; border-bottom: 1px solid #d9d9d9;\n    display: inline-block;\n    color: #999;\n    padding: 2px;">NLP民工的乐园</div>\n</center>\n<br><br><br><br>\n\n[![](https://img.shields.io/github/stars/fighting41love/funnlp?style=social)](https://github.com/fighting41love/funnlp)\n[![](https://img.shields.io/badge/dynamic/json?color=blue&label=%E7%9F%A5%E4%B9%8E%E5%85%B3%E6%B3%A8&query=%24.data.totalSubs&url=https%3A%2F%2Fapi.spencerwoo.com%2Fsubstats%2F%3Fsource%3Dzhihu%26queryKey%3Dmountain-blue-64)](https://www.zhihu.com/people/mountain-blue-64)\n[![](data/.logo图片/.捐赠图片/.Citations-487-red.svg)](https://scholar.google.com/citations?hl=en&user=aqZdfDUAAAAJ)\n\n[![](data/.logo图片/.捐赠图片/.Home-%E4%BA%BA%E7%94%9F%E6%B5%AA%E8%B4%B9%E6%8C%87%E5%8D%97-brightgreen.svg)](https://scholar.go

### Takeaway
- The readme acquired contains foreign language, we will going to drop those languages that is not in English, due to we acquired about 490 rows, we met the requirement of at least 100 rows.

#### Install the package needed to detect the language
- \# pip install langdetect

### Key takeaway
- The most starred README data on github was collected on October 17, 2022
- The data has 490 rows
- We found that the readme contains foreign language, therefore, we downloaded a langdetect package for further wrangle the dataset
- we also spot some 'none' value in our language, we will do further wrangle with that data as well
- Following this acquire, we are going to prepare for our exploration

In [31]:
# clean the data acquired
# We are acquire the data that is cleaned up with tokenized, stemmed, and lemmatized
# add those columns into the dataframe and create a final data frame
df=prepare.wrangle_data()

In [34]:
# we will going to drop 'none' value in our language column
df=df.dropna()
df

repo    language  \
2                  DataScienceSpecialization/courses        HTML   
3                              fighting41love/funNLP      Python   
4                                   magento/magento2         PHP   
5                                    arduino/Arduino        Java   
6                                     elunez/eladmin        Java   
..                                               ...         ...   
484                   harshalbenake/hbworkspace1-100        Java   
485                                   rprokap/pset-9  JavaScript   
486                                 konzy/mass_clone       Shell   
487                               biter777/countries          Go   
488  sayantann11/all-classification-templetes-for-ML      Python   

                                       readme_contents  \
2    \n### Data Science Specialization\n\nThese are...   
3    <center>\n    <img style="border-radius: 0.312...   
4    \n<p align="center">\n<a href="https://www.cod...   
5    <p align="center">\n\t<img src="http://content...   
6    <h1 style="text-align: center">ELADMIN 后台管理系统<...   
..                                                 ...   
484  # hbworkspace1-100\n\n(1) Name :- accelormeter...   
485  # pset-9\n      CREDITS SEQUENCE              ...   
486  # mass_clone\nThis is a shell script that will...   
487  # countries\r\n\r\nCountries - ISO 3166 (ISO31...   
488  # all-classification-templetes-for-ML\nClassif...   

                                               stemmed  \
2    data scienc specializationthes cours materi jo...   
3    nlpdatalogocitations487redsvgdatalogohomee4bab...   
4    magento open sourcewelcom magento open sourc p...   
5    &#9; arduino opensourc physic comput platform ...   
6    eladmin aurstargithub starsgithub fork spring ...   
..                                                 ...   
484  hbworkspace11001 name accelormetersensordescri...   
485  pset9credit sequencenewspap headlin montagehea...   
486  massclonethi shell script clone multipl reposi...   
487  countriescountri iso 3166 iso31661 iso3166 dig...   
488  allclassificationtempletesformlclassif machin ...   

                                            lemmatized  
2    data science specializationthese course materi...  
3    nlpdatalogocitations487redsvgdatalogohomee4bab...  
4    magento open sourcewelcome magento open source...  
5    &#9; arduino opensource physical computing pla...  
6    eladmin aurstargithub starsgithub fork spring ...  
..                                                 ...  
484  hbworkspace11001 name accelormetersensordescri...  
485  pset9credits sequencenewspaper headline montag...  
486  massclonethis shell script clone multiple repo...  
487  countriescountries iso 3166 iso31661 iso3166 d...  
488  allclassificationtempletesformlclassification ...  

[365 rows x 5 columns]

In [42]:
# we will going to use the installed package 
# to filter out the readme contents that is in English only

# we are going to import a new libary for this
import langdetect as ld

# we created a new function to detect the non-english language in read me
# the function will return the result when it is in english,, elso will not return the result
def is_en(txt):
    try:
        return ld.detect(txt)=='en'
    except:
        return False

# we applied the function we created 
df = df[df['readme_contents'].apply(is_en)]

In [45]:
# check how many 
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 291 entries, 2 to 488
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   repo             291 non-null    object
 1   language         291 non-null    object
 2   readme_contents  291 non-null    object
 3   stemmed          291 non-null    object
 4   lemmatized       291 non-null    object
dtypes: object(5)
memory usage: 13.6+ KB


# Overall Wrangle Takeaway
- The data aquired on October 17th, 2022 with 490 rows
- The data contains 'none' value in the language column, we dropped all 'none' values 
- The data contains foreign language in the readme contents column, we drop all those non-english values
- We finalized with 291 rows, and ready for exploration

# Exploration

- our dataframe should contains numerical data for better exploration, therefore, we are going to created some columns that able to represent the overall struture of the dataframe. we picked df.lemmatized

In [56]:
df=explore.feature_engineering(df)

repo    language  \
2                  DataScienceSpecialization/courses        HTML   
4                                   magento/magento2         PHP   
5                                    arduino/Arduino        Java   
9         bloominstituteoftechnology/Preprocessing-I         CSS   
11                                  magento/magento2         PHP   
..                                               ...         ...   
484                   harshalbenake/hbworkspace1-100        Java   
485                                   rprokap/pset-9  JavaScript   
486                                 konzy/mass_clone       Shell   
487                               biter777/countries          Go   
488  sayantann11/all-classification-templetes-for-ML      Python   

                                       readme_contents  \
2    \n### Data Science Specialization\n\nThese are...   
4    \n<p align="center">\n<a href="https://www.cod...   
5    <p align="center">\n\t<img src="http://content...   
9    # Preprocessing I: Digital Resume\n\nFor this ...   
11   \n<p align="center">\n<a href="https://www.cod...   
..                                                 ...   
484  # hbworkspace1-100\n\n(1) Name :- accelormeter...   
485  # pset-9\n      CREDITS SEQUENCE              ...   
486  # mass_clone\nThis is a shell script that will...   
487  # countries\r\n\r\nCountries - ISO 3166 (ISO31...   
488  # all-classification-templetes-for-ML\nClassif...   

                                               stemmed  \
2    data scienc specializationthes cours materi jo...   
4    magento open sourcewelcom magento open sourc p...   
5    &#9; arduino opensourc physic comput platform ...   
9    preprocess digit resumefor project build digit...   
11   magento open sourcewelcom magento open sourc p...   
..                                                 ...   
484  hbworkspace11001 name accelormetersensordescri...   
485  pset9credit sequencenewspap headlin montagehea...   
486  massclonethi shell script clone multipl reposi...   
487  countriescountri iso 3166 iso31661 iso3166 dig...   
488  allclassificationtempletesformlclassif machin ...   

                                            lemmatized  \
2    data science specializationthese course materi...   
4    magento open sourcewelcome magento open source...   
5    &#9; arduino opensource physical computing pla...   
9    preprocessing digital resumefor project buildi...   
11   magento open sourcewelcome magento open source...   
..                                                 ...   
484  hbworkspace11001 name accelormetersensordescri...   
485  pset9credits sequencenewspaper headline montag...   
486  massclonethis shell script clone multiple repo...   
487  countriescountries iso 3166 iso31661 iso3166 d...   
488  allclassificationtempletesformlclassification ...   

                                             word_list  \
2    [data, science, specializationthese, course, m...   
4    [magento, open, sourcewelcome, magento, open, ...   
5    [9, arduino, opensource, physical, computing, ...   
9    [preprocessing, digital, resumefor, project, b...   
11   [magento, open, sourcewelcome, magento, open, ...   
..                                                 ...   
484  [hbworkspace11001, name, accelormetersensordes...   
485  [pset9credits, sequencenewspaper, headline, mo...   
486  [massclonethis, shell, script, clone, multiple...   
487  [countriescountries, iso, 3166, iso31661, iso3...   
488  [allclassificationtempletesformlclassification...   

                                          unique_words  \
2    [data, science, specializationthese, course, m...   
4    [magento, open, sourcewelcome, source, project...   
5    [9, arduino, opensource, physical, computing, ...   
9    [preprocessing, digital, resumefor, project, b...   
11   [magento, open, sourcewelcome, source, project...   
..                                                 ...   
484  [hbworkspace11001, name, accelor